File: oclfile.py
import ocl
import math
import os
import pickle

from enum import Enum

def free(x):
  del x


class OclFile:
  oclfile_instances = []
  oclfile_index = dict({})

  def __init__(self):
    self.name = ""
    self.position = 0
    self.actualFile = None
    self.lastRead = None
    self.eof = False
    OclFile.oclfile_instances.append(self)

  def newOclFile(nme) : 
    f = createByPKOclFile(nme)
    return f

  def newOclFile_Read(f) : 
    f.openRead()
    return f

  def newOclFile_Write(f) : 
    f.openWrite()
    return f

  def newOclFile_ReadB(f) : 
    f.openReadB()
    return f

  def newOclFile_WriteB(f) : 
    f.openWriteB()
    return f

  def getName(self) : 
    return self.name

  def compareTo(self,f) : 
    if self.name < f.name : 
      return -1 
    if self.name > f.name : 
      return 1
    return 0

  def canRead(self) : 
    if self.actualFile != None : 
      return self.actualFile.readable()
    return False

  def canWrite(self) : 
    if self.actualFile != None : 
      return self.actualFile.writable()
    return False


  def openWrite(self) : 
    if self.name == "System.out" or self.name == "System.err" : 
      pass 
    else : 
      self.actualFile = open(self.name, 'w')
    self.position = 0

  def openRead(self) : 
    if self.name == "System.in" : 
      pass
    else : 
      self.actualFile = open(self.name, 'r')
    self.position = 0

  def openWriteB(self) : 
    if self.name == "System.out" or self.name == "System.err" : 
      pass 
    else : 
      self.actualFile = open(self.name, 'wb')
    self.position = 0

  def openReadB(self) : 
    if self.name == "System.in" : 
      pass
    else : 
      self.actualFile = open(self.name, 'rb')
    self.position = 0

  def exists(self) : 
    return os.path.isfile(self.name)

  def isFile(self) : 
    return os.path.isfile(self.name)

  def isDirectory(self) : 
    return os.path.isdir(self.name)

  def isAbsolute(self) : 
    return os.path.isabs(self.name)

  def getAbsolutePath(self) : 
    return os.path.abspath(self.name)

  def getPath(self) : 
    return os.path.realpath(self.name)

  def getParent(self) : 
    pth = os.path.abspath(self.name)
    return os.path.dirname(pth)

  def getParentFile(self) : 
    pth = os.path.abspath(self.name)
    str = os.path.dirname(pth)
    if str != None : 
      return OclFile.newOclFile(str)

  def lastModified(self) : 
    return int(os.path.getmtime(self.name))

  def length(self) : 
    return os.path.getsize(self.name)

  def delete(self) : 
    os.remove("./" + self.name)
    if os.path.isfile(self.name) : 
      return False
    return True

  def list(self) : 
    return os.listdir()

  def listFiles(self) : 
    sq = os.listdir()
    res = []
    for x in sq : 
      res.append(OclFile.newOclFile(x))
    return res

  def mkdir(self) : 
    try: 
      os.mkdir(self.name)
    except OSError as error: 
      return False
    return True

  def print(self, s) :
    if self.name == "System.out" or self.name == "System.err" : 
      print(str(s), end="")
    else : 
      self.write(s)

  def println(self, s) :
    if self.name == "System.out" or self.name == "System.err" : 
      print(str(s))
    else : 
      self.writeln(s)

  def printf(self, f, sq) :
    print(f % tuple(sq), end="")

  def write(self, s) :
    if self.actualFile != None : 
      self.actualFile.write(s)

  def writeObject(self, s) :
    if self.actualFile != None : 
      pickle.dump(s,self.actualFile)

  def writeln(self, s) :
    if self.actualFile != None : 
      self.actualFile.write(s + "\n")

  def flush(self) :
    if self.actualFile != None : 
      self.actualFile.flush()

  def hasNext(self) : 
    if self.name == "System.in" :
      if self.eof == True : 
        return False 
      try : 
        s = input("")
        self.lastRead = s
        return True
      except EOFError : 
        self.eof = True
        self.lastRead = None
        return False
    return False 
    

  def read(self) :
    if self.actualFile != None :
      try :  
        s = self.actualFile.read(1)
        self.lastRead = s
        return s
      except EOFError : 
        self.eof = True
        self.lastRead = None
        return None
    if self.name == "System.in" : 
      if self.lastRead != None : 
        return self.lastRead
      try : 
        s = input("")
        self.lastRead = s
        return s
      except EOFError : 
        self.eof = True
        self.lastRead = None
        return None
    return ""

  def getCurrent(self) : 
    return self.lastRead
      
  def readObject(self) :
    if self.actualFile != None : 
      return pickle.load(self.actualFile)
    return None

  def readLine(self) :
    if self.actualFile != None : 
      return self.actualFile.readline()
    if self.name == "System.in" : 
      s = input("")
      return s
    return ""

  def readAll(self) :
    if self.actualFile != None : 
      return self.actualFile.read()
    return ""

  def mark(self) : 
    if self.actualFile != None : 
      self.position = self.actualFile.tell()
    else : 
      self.position = 0

  def reset(self) : 
    if self.actualFile != None : 
      self.actualFile.seek(self.position)
    
  def skipBytes(self,n) : 
    if self.actualFile != None : 
      self.position = self.actualFile.tell()
      self.position = self.position + n
      self.actualFile.seek(self.position)


  def closeFile(self) : 
    if self.actualFile != None : 
      self.actualFile.close()
      self.actualFile = None


  def killOclFile(oclfile_x) :
    oclfile_instances = ocl.excludingSet(oclfile_instances, oclfile_x)
    free(oclfile_x)

def createOclFile():
  oclfile = OclFile()
  return oclfile

def allInstances_OclFile():
  return OclFile.oclfile_instances

def getOclFileByPK(_ex) :
  if (_ex in OclFile.oclfile_index) :
    return OclFile.oclfile_index[_ex]
  else :
    return None


def getOclFileByPKs(_exs) :
  result = []
  for _ex in _exs :
    if (_ex in OclFile.oclfile_index) :
      result.append(OclFile.oclfile_index[_ex])
  return result


def createByPKOclFile(_value):
  result = getOclFileByPK(_value)
  if (result != None) :
    return result
  else :
    result = OclFile()
    result.name = _value
    OclFile.oclfile_index[_value] = result
    return result


# f1 = OclFile.newOclFile("./subdir")
# f2 = OclFile.newOclFile_Read(f1)
# print(f2.length())
# print(f2.isFile())
# print(f2.isDirectory())
# print(f2.lastModified())
# print(f2.isAbsolute())

# f2.closeFile()
# print(f1.mkdir())

# print(f2.canRead())

# f = OclFile.newOclFile("f.txt")
# print(f.getAbsolutePath())
# print(f.getParent())
# d = f.getParentFile()
# print(d.getName())
# print(f.getPath())


# f = open("w.txt", 'w')
# f.write("some text")
# f.write("more text")
# f.flush()
# f.close()

System_in = OclFile.newOclFile("System.in")
System_out = OclFile.newOclFile("System.out")
System_err = OclFile.newOclFile("System.err")

# f = OclFile.newOclFile("obj.dat")
# fx = OclFile.newOclFile_ReadB(f)

# class Person : 
#   def __init__(self,nme) : 
#     self.name = nme

# p1 = Person("Hamlet")
# p2 = Person("Horatio")

# p1 = fx.readObject()
# p2 = fx.readObject()
# fx.closeFile()

# print(p1.name)
# print(p2.name)

# System_out.printf("%d %s", [120, "degrees"])

# at end of file - go to end of file, find the 
# position there, return to current position. 
# EOF if they are the same. 
# pos = fh.tell()
# fh.seek(0, 2)
# file_size = fh.tell()
# if pos == file_size : 
#   return True
# fh.seek(pos)
# return False

# sq = []
# while System_in.hasNext() : 
#   x = System_in.getCurrent()
#   sq.append(x)
# print(sq)

File: libraries/OclFile.java
import java.io.*; 
import java.nio.file.Files; 
import java.nio.file.Path;
import java.util.Map; 
import java.util.HashMap;
import java.util.Collection;
import java.util.List;
import java.util.ArrayList;
import java.util.Set;
import java.util.HashSet;
import java.util.TreeSet;
import java.util.Collections;

class OclFile { 
  static 
  { OclFile.OclFile_index = new HashMap<String,OclFile>();
    OclFile systemIn = OclFile.createByPKOclFile("System.in"); 
    systemIn.inputStream = System.in; 
    systemIn.inputStreamReader = 
          new BufferedReader(
            new InputStreamReader(System.in)); 

    OclFile systemOut = OclFile.createByPKOclFile("System.out"); 
    systemOut.outputStream = System.out; 
    OclFile systemErr = OclFile.createByPKOclFile("System.err"); 
    systemErr.outputStream = System.err; 
  } 

  public File actualFile = null; 
  public InputStream inputStream = null; 
  public OutputStream outputStream = null; 
  public Reader inputStreamReader = null; 
  public Writer outputStreamWriter = null; 

  private long position = 0L; 
  private long markedPosition = 0L; 
  private boolean eof = false; 

  String name = ""; /* primary */
  static Map<String,OclFile> OclFile_index;


  OclFile() { }

  OclFile(String nme)
  { name = nme; 
    OclFile.OclFile_index.put(name,this);
  } 

  OclFile(File c) 
  { this(c.getName());
    actualFile = c; 
  }


  static OclFile createOclFile() 
  { OclFile result = new OclFile();
    return result; 
  }


  static OclFile createByPKOclFile(String namex)
  { OclFile result = OclFile.OclFile_index.get(namex);
    if (result != null) 
    { return result; }
    result = new OclFile();
    OclFile.OclFile_index.put(namex,result);
    result.name = namex;
    return result; 
  }

  static OclFile getOclFileByPK(String namex)
  { OclFile result = OclFile.OclFile_index.get(namex);
    if (result != null) 
    { return result; }
    result = new OclFile();
    OclFile.OclFile_index.put(namex,result);
    result.name = namex;
    return result; 
  }

  static void killOclFile(String namex)
  { OclFile rem = OclFile_index.get(namex);
    if (rem == null) { return; }
    ArrayList<OclFile> remd = new ArrayList<OclFile>();
    remd.add(rem);
    OclFile_index.remove(namex);
  }

  public static OclFile newOclFile(String nme)
  { OclFile res = createByPKOclFile(nme); 
    res.actualFile = new File(nme); 
    return res; 
  } 

  public static OclFile newOclFile_Read(OclFile f)
  { /* System.out.println("Creating file for " + f.inputStream + " " + f.inputStreamReader); */ 

    if (f.inputStream == null) // create stream from file 
    { try 
      { f.inputStream = new FileInputStream(f.actualFile); 
        f.inputStreamReader = 
            new BufferedReader(new FileReader(f.actualFile));
        f.inputStream.mark((int) f.length());
        f.position = 0;   
      } 
      catch (Exception e) { } 
      return f; 
    }
    else if (f.inputStreamReader == null) // reader from stream
    { try { 
        f.inputStreamReader = 
          new BufferedReader(
            new InputStreamReader(f.inputStream)); 
        f.position = 0;
        return f; 
      } catch (Exception _ex) { _ex.printStackTrace(); }
    }  
    return f;
  } 

  // public void openRead() is the same

  public static OclFile newOclFile_Write(OclFile f)
  { if (f.outputStream == null) 
    { try 
      { f.outputStream = new FileOutputStream(f.actualFile);
        f.outputStreamWriter = new BufferedWriter(new FileWriter(f.actualFile));
        f.position = 0;  
      } 
      catch (Exception e) 
      { e.printStackTrace(); }
    }   
    else if (f.outputStreamWriter == null) // writer from stream
    { try { 
        f.outputStreamWriter = 
          new BufferedWriter(
            new OutputStreamWriter(f.outputStream)); 
        f.position = 0;
      } catch (Exception _ex) { }
    }  
    return f;
  }

  // public void openWrite() is the same

  public static OclFile newOclFile_RW(OclFile f)
  { if (f.inputStream == null) // create stream from file 
    { try 
      { f.inputStream = new FileInputStream(f.actualFile); 
        f.inputStreamReader = 
            new BufferedReader(new FileReader(f.actualFile));
        f.inputStream.mark((int) f.length());
        f.position = 0;   
      } 
      catch (Exception e) { } 
    }
    else if (f.inputStreamReader == null) // reader from stream
    { try { 
        f.inputStreamReader = 
          new BufferedReader(
            new InputStreamReader(f.inputStream)); 
        f.position = 0;
      } catch (Exception _ex) { _ex.printStackTrace(); }
    }  

    if (f.outputStream == null) 
    { try 
      { f.outputStream = new FileOutputStream(f.actualFile);
        f.outputStreamWriter = new BufferedWriter(new FileWriter(f.actualFile));
        f.position = 0;  
      } 
      catch (Exception e) 
      { e.printStackTrace(); }
    }   
    else if (f.outputStreamWriter == null) // writer from stream
    { try { 
        f.outputStreamWriter = 
          new BufferedWriter(
            new OutputStreamWriter(f.outputStream)); 
        f.position = 0;
      } catch (Exception _ex) { }
    }  

    return f;
  }

  public static OclFile newOclFile_ReadB(OclFile f)
  { if (f.inputStream == null && f.actualFile != null) 
    { try 
      { f.inputStream = new ObjectInputStream(new FileInputStream(f.actualFile)); 
        f.position = 0; 
      } 
      catch (Exception e) { } 
      return f; 
    }
    
    if (f.inputStream != null && f.actualFile == null) 
    { try 
      { f.inputStream = new ObjectInputStream(f.inputStream); 
        f.position = 0; 
      } 
      catch (Exception e) { } 
      return f; 
    }

    return f;
  } 

  // public void openReadB() is the same

  public static OclFile newOclFile_WriteB(OclFile f)
  { if (f.outputStream == null && f.actualFile != null) 
    { try 
      { f.outputStream = 
          new ObjectOutputStream(
            new FileOutputStream(f.actualFile)); 
      } 
      catch (Exception e) { }  
      return f; 
    }

    if (f.outputStream != null && f.actualFile == null) 
    { try 
      { f.outputStream = 
             new ObjectOutputStream(f.outputStream); 
        f.position = 0; 
      } 
      catch (Exception e) { } 
      return f; 
    }

    return f;
  }

  // public void openWriteB() is the same

  public void setFile(File f)
  { actualFile = f; } 

  public String getName()
  { return name; }

  public int compareTo(OclFile f)
  { return name.compareTo(f.getName()); }


  public boolean canRead()
  { if (actualFile != null) 
    { return actualFile.canRead(); } 
    return false;
  }

  public boolean canWrite()
  { if (actualFile != null) 
    { return actualFile.canRead(); } 
    return false;
  }

  public boolean isOpen()
  { if (outputStream == null && 
        outputStreamWriter == null && 
        inputStream == null && 
        inputStreamReader == null) 
    { return false; } 
    return true; 
  } 

  public void closeFile()
  { if (outputStreamWriter != null) 
    { try { outputStreamWriter.close(); } 
      catch (Exception _x) { } 
    } 
    if (inputStreamReader != null) 
    { try { inputStreamReader.close(); } 
      catch (Exception _x) { } 
    } 

    outputStreamWriter = null; 
    inputStreamReader = null; 

    if (outputStream != null)
    { try { outputStream.close(); }
      catch (Exception _e) { } 
    } 
    if (inputStream != null)
    { try { inputStream.close(); }
      catch (Exception _f) { } 
    } 

    inputStream = null; 
    outputStream = null; 

    position = 0; 
  }

  public boolean exists()
  { if (actualFile != null) 
    { return actualFile.exists(); }
    return false; 
  } 

  public boolean isFile()
  { if (actualFile != null) 
    { return actualFile.isFile(); }
    return false; 
  } 

  public boolean isDirectory()
  { if (actualFile != null) 
    { return actualFile.isDirectory(); }
    return false; 
  } 

  public boolean isHidden()
  { if (actualFile != null) 
    { return actualFile.isHidden(); }
    return false; 
  } 

  public void setHidden(boolean b)
  { if (actualFile != null) 
    { try { Files.setAttribute(actualFile.toPath(), "dos:hidden", b); }
      catch (Exception _io) { } 
    } 
  } 

  public void setExecutable(boolean h)
  { if (actualFile != null) 
    { actualFile.setExecutable(h); }
  } 

  public boolean isExecutable()
  { if (actualFile != null) 
    { return actualFile.getName().endsWith(".exe"); }
    return false; 
  } 
  
  public void setReadOnly()
  { if (actualFile != null) 
    { actualFile.setReadOnly(); }
  } 

  public boolean isAbsolute()
  { if (actualFile != null) 
    { return actualFile.isAbsolute(); }
    return false; 
  }


  public String getAbsolutePath()
  { if (actualFile != null) 
    { return actualFile.getAbsolutePath(); }
    return name; 
  }

  public String getPath()
  { if (actualFile != null) 
    { return actualFile.getPath(); }
    return name;
  }

  public String getParent()
  { if (actualFile != null) 
    { return actualFile.getParent(); } 
    return ""; 
  } 

  public OclFile getParentFile()
  { if (actualFile != null && 
        actualFile.getParentFile() != null) 
    { OclFile result = 
         new OclFile(actualFile.getParentFile());
      return result; 
    }  
    OclFile f = new OclFile(getParent());
    return f;
  }

  public long lastModified()
  { if (actualFile != null) 
    { return actualFile.lastModified(); } 
    return 0; 
  } 

  public long length()
  { if (actualFile != null) 
    { return actualFile.length(); } 
    return 0; 
  } 

  public boolean delete()
  { if (actualFile != null) 
    { return actualFile.delete(); }
    return false; 
  } // and close everything

  public static boolean deleteFile(String nme)
  { File fle = new File(nme); 
    if (fle != null) 
    { return fle.delete(); }
    return false; 
  } 
  
  public static boolean renameFile(String from, String to)
  { File f1 = new File(from); 
    File f2 = new File(to); 
	if (f1 != null && f2 != null)
	{ Path p1 = f1.toPath(); 
	  Path p2 = f2.toPath(); 
	  try { Files.move(p1,p2); } catch (Exception _m) { return false; }
	  return true; 
	} 
	return false; 
  } 
  
  public static OclFile createTemporaryFile(String f, String ext)
  { File f1 = new File(f + "." + ext); 
    if (f1 == null) { return null; }
    Path p1 = f1.toPath();
    try { Files.createFile(p1); } 
	catch (Exception _ex) { return null; }
    OclFile res = createByPKOclFile(f + "." + ext); 
	res.actualFile = f1; 
	return res; 
  } 
   	
  public ArrayList<String> list()
  { ArrayList<String> result = new ArrayList<String>();
    if (actualFile != null) 
    { String[] fnames = actualFile.list();
      if (fnames == null) 
      { return result; } 
      for (int i = 0; i < fnames.length; i++)
      { result.add(fnames[i]); }
      return result;
    }
    return result;
  }

  public ArrayList<OclFile> listFiles()
  { ArrayList<OclFile> result = new ArrayList<OclFile>();
    if (actualFile != null) 
    { File[] fnames = actualFile.listFiles();
      for (int i = 0; i < fnames.length; i++)
      { result.add(new OclFile(fnames[i])); }
      return result;
    }
    return result;
  }

  public boolean mkdir()
  { boolean result = false;
    if (actualFile != null)
    { return actualFile.mkdir(); }
    return result;
  }

  public boolean mkdirs()
  { boolean result = false;
    if (actualFile != null)
    { return actualFile.mkdirs(); }
    return result;
  }

  public void write(String s)
  { 
    if (outputStreamWriter != null)
    { try { outputStreamWriter.write(s); }
      catch (Exception _e) 
      { _e.printStackTrace(); } 
    } 
  }

  public OclFile append(String s)
  { write(s); 
    return this; 
  }

  public void writeN(ArrayList<String> sq, int n)
  { 
    if (outputStreamWriter != null)
    { try { 
        for (int i = 0; i < n && i < sq.size(); i++) 
        { String s = sq.get(i); 
          outputStreamWriter.write(s); 
        }
      } 
      catch (Exception _e) 
      { _e.printStackTrace(); } 
    } 
  }

  public void writeByte(int c)
  { 
    if (outputStreamWriter != null)
    { try { outputStreamWriter.write(c); }
      catch (Exception _e) 
      { _e.printStackTrace(); } 
    } 
  }

  public void writeNbytes(ArrayList<Integer> sq, int n)
  { 
    if (outputStreamWriter != null)
    { try { 
        for (int i = 0; i < n && i < sq.size(); i++) 
        { int x = sq.get(i); 
          outputStreamWriter.write(x); 
        }
      } 
      catch (Exception _e) 
      { _e.printStackTrace(); } 
    } 
  }

  public void writeln(String s)
  {
    if (outputStreamWriter != null)
    { try { outputStreamWriter.write(s + "\n"); }
      catch (Exception _e) { } 
    } 
  }

  public void writeObject(Object obj)
  { if (outputStream != null && 
        outputStream instanceof ObjectOutputStream) 
    { try { ((ObjectOutputStream) outputStream).writeObject(obj); }
      catch (Exception e) 
      { e.printStackTrace(); } 
    }  
    else if (outputStream != null) 
    { try { 
        ObjectOutputStream objstream = 
          new ObjectOutputStream(outputStream); 
        objstream.writeObject(obj); 
      }
      catch (Exception e) 
      { e.printStackTrace(); } 
    }  

    
  } 

  public void writeMap(Map m)
  { writeObject(m); } 

  public Object readObject()
  { if (inputStream != null && 
        inputStream instanceof ObjectInputStream) 
    { try 
      { Object res = 
             ((ObjectInputStream) inputStream).readObject(); 
        return res; 
      }
      catch (Exception e) 
      { eof = true; } 
    } 
    else if (inputStream != null) 
    { try { 
        ObjectInputStream objstream = 
          new ObjectInputStream(inputStream); 
        Object res = objstream.readObject();
        return res;  
      }
      catch (Exception e) 
      { e.printStackTrace(); } 
    }  


    return null;  
  } 

  public Map readMap()
  { Object res = readObject(); 
    if (res instanceof Map) 
    { return (Map) res; } 
    return new HashMap(); 
  } 


  public String readLine()
  { if (inputStreamReader != null && 
        inputStreamReader instanceof BufferedReader) 
    { try 
      { 
        String ss = ((BufferedReader) inputStreamReader).readLine(); 
        position += ss.length() + 1; 
        return ss; 
      } catch (Exception _e) { 
          _e.printStackTrace(); 
          eof = true; 
          return null; 
      } 
    } 

    return ""; 
  } 

  public String read()
  { if (inputStreamReader != null) 
    { try 
      { 
        int ch = inputStreamReader.read();
        position++;
        if (ch == -1) 
        { eof = true; } 
        return Ocl.byte2char(ch);   
      } catch (Exception _e) 
        { eof = true; } 
    } 
 
    return ""; 
  } 

  public ArrayList<String> readN(int n)
  { ArrayList<String> res = new ArrayList<String>(); 
    if (inputStreamReader != null) 
    { try 
      { int ind = 0; 
        int ch = inputStreamReader.read();
        position++;
        while (ch != -1 && ind < n)
        { res.add(Ocl.byte2char(ch)); 
          ch = inputStreamReader.read();
          position++;
          ind++; 
        } 
        return res;   
      } catch (Exception _e) 
        { eof = true; } 
    } 
    return res; 
  } 

  public int readByte()
  { if (inputStreamReader != null) 
    { try 
      { 
        int ch = inputStreamReader.read();
        position++;
        return ch;   
      } catch (Exception _e) 
        { eof = true; } 
    } 
 
    return -1; 
  } 

  public ArrayList<Integer> readNbytes(int n)
  { ArrayList<Integer> res = new ArrayList<Integer>(); 
    if (inputStreamReader != null) 
    { try 
      { int ind = 0; 
        int ch = inputStreamReader.read();
        position++;
        while (ch != -1 && ind < n)
        { res.add(ch); 
          ch = inputStreamReader.read();
          position++;
          ind++; 
        } 
        return res;   
      } catch (Exception _e) 
        { eof = true; } 
    } 
    return res; 
  } 

  public ArrayList<Integer> readAllBytes()
  { ArrayList<Integer> res = new ArrayList<Integer>(); 
    if (inputStreamReader != null) 
    { try 
      { 
        int ch = inputStreamReader.read();
        position++;
        while (ch != -1)
        { res.add(ch); 
          ch = inputStreamReader.read();
          position++;
        } 
        return res;   
      } catch (Exception _e) 
        { eof = true; } 
    } 
    return res; 
  } 

  public String readAll()
  { if (inputStreamReader != null && 
        inputStreamReader instanceof BufferedReader) 
    { StringBuffer br = new StringBuffer(); 
      try 
      { String lne = 
          ((BufferedReader) inputStreamReader).readLine();
        while (lne != null) 
        { br.append(lne + "\n"); 
          lne = 
            ((BufferedReader) inputStreamReader).readLine();
        } 
        position = length(); 
      } catch (Exception _e) {}
      return br.toString();    
    } 
 
    return null; 
  } 

  public ArrayList<String> readAllLines()
  { ArrayList<String> res = new ArrayList<String>(); 

    if (inputStreamReader != null && 
        inputStreamReader instanceof BufferedReader) 
    { StringBuffer br = new StringBuffer(); 
      try 
      { String lne = 
          ((BufferedReader) inputStreamReader).readLine();
        while (lne != null) 
        { res.add(lne); 
          lne = 
            ((BufferedReader) inputStreamReader).readLine();
        } 
        position = length(); 
      } catch (Exception _e) {}
      return res;    
    } 
 
    return res; 
  } 

  public static void copyFromTo(OclFile source, OclFile target)
  { if (source == null || target == null)
    { return; } 

    if (source.inputStreamReader != null) 
    { try 
      { 
        int ch = source.inputStreamReader.read();
        source.position++;
        while (ch != -1)
        { target.writeByte(ch); 
          ch = source.inputStreamReader.read();
          source.position++;
        } 
        
      } catch (Exception _e) 
        { source.eof = true; } 
    } 
  } 
        

  public void flush()
  {
    if (outputStream != null)
    { try { outputStream.flush(); }
      catch (Exception _e) { } 
    } 
    
    if (outputStreamWriter != null) 
    { try { outputStreamWriter.flush(); } 
      catch (Exception _x) { } 
    } 
  }

  public void print(String s)
  { if ("System.out".equals(name))
    { System.out.print(s); } 
    else if ("System.err".equals(name))
    { System.err.print(s); }
    else
    { write(s); }
  }

  public void println(String s)
  { if ("System.out".equals(name))
    { System.out.println(s); } 
    else if ("System.err".equals(name))
    { System.err.println(s); }
    else
    { writeln(s); }
  }

  public void printf(String s, ArrayList sq)
  { Object[] args = new Object[sq.size()]; 
    for (int i = 0; i < sq.size(); i++) 
    { args[i] = sq.get(i); }
    String formattedString = String.format(s,args);  

    if ("System.out".equals(name))
    { System.out.printf(s,args); } 
    else if ("System.err".equals(name))
    { System.err.printf(s,args); }
    else 
    { write(formattedString); }
  }


  public void mark()
  { if (inputStream != null)
    { inputStream.mark((int) length()); 
      markedPosition = position; 
    }
  } 
        
  public void reset()
  { if (inputStream != null)
    { try { inputStream.reset();
            position = markedPosition;
      }
      catch (Exception _e) { } 
    } 
  } 

  public void skipBytes(int n)
  {
    if (inputStream != null)
    { try { inputStream.skip(n); 
            position += n; 
      }
      catch (Exception _e) 
      { eof = true; } 
    }  
  } 

  public void setPosition(long pos)
  { if (inputStream != null)
    { try { inputStream.reset(); // assume, to 0
            position = pos;
            skipBytes((int) pos); 
      }
      catch (Exception _e) 
      { eof = true; } 
    } 
  } 

  public boolean getEof() 
  { return eof; } 

  public long getPosition()
  { return position; } 

  protected void finalize()
  { closeFile(); } 

 /* public static void main(String[] args)
  { OclFile ff = OclFile.newOclFile("ff.txt"); 
    OclFile gg = OclFile.newOclFile_Read(ff); 
    ArrayList<Integer> ss = gg.readNbytes(4);
    System.out.println(ss); 
    // gg.skipBytes(20); 
    ArrayList<String> cc = gg.readN(4);
    cc = gg.readN(4); 
    System.out.println(cc); 
    gg.closeFile(); 
  }  */ 

}

